storage/ast/node_convert.go

```diff
@@ -70,6 +70,65 @@ func toFloat64Nodes(nodes ...Node) ([]Float64Node, bool) {
 	return result, true
 }
 
+type boolBinaryOp int
+
+const (
+	AndOp boolBinaryOp = iota
+	OrOp
+)
+
+type BooleanLogicExprNode struct {
+	left  Node
+	right Node
+	op    boolBinaryOp
+}
+
```

Comment:
Adding a temporary conversion node here. And/Or requires that the left/right be bools, however there may be some types which resolve to bool, but don't start that way. This allows us to defer that check until child nodes have reached their final state.